### PR TITLE
useInterpret()/useMachine() no longer creates an empty machine.context object when it was set as undefined.

### LINKED ISSUE #2084: Inconsistent behavior for initial state using `interpret` vs. `useInterpret`
**Description**
Using `interpret` and `useInterpret` results in different runtime behavior.
If a machine **without initial context** is started using `interpret` the `context` resolves to `undefined`, if it is started with `useInterpret` the context resolves to `{}`, this in turn affects the behavior of `assign(...)`.
Because of the huge advantage of xstate being framework independent, I think that's unwanted or even potential risky behavior.
This inconsistency can easily be solved by simply providing an initial context, but **it currently fails silently if one switches from `useInterpret` to `interpret`**.

The reason for this is: (@VanTanev 🙏)
https://github.com/davidkpiano/xstate/blob/14cdef612df29b9e067a34ca96b0750e729c9aff/packages/xstate-react/src/useInterpret.ts#L90-L93

> [@VanTanev on discord](https://discord.com/channels/795785288994652170/799416943324823592/831444212322533418): `useInterpret()` calls `machine.withConfig()`. The second argument to `withConfig()` is a new `context` object, that overwrites `context: undefined` in the original machine definition. Here it's kind of accidental, and possibly **a more consistent way to write this would be something like**:

```ts
const machineWithConfig = machine.withConfig(machineConfig, (machine.context || context ? {
  ...machine.context,
  ...context
} : undefined) as TContext);
```

**Expected Result**
Consistent runtime behavior using either `interpret` or `useInterpret`. Preferably the behavior of `interpret`, because it is the underlying framework-agnostic version of `useInterpret`. _The power of writing framework-agnostic logic code is just too valuable to be given up 😍._

**Actual Result**
Different runtime behavior (see **Reproduction**).

**Reproduction**
_xstate 4.17.1; @xstate/inspect 0.4.1; @xstate/react 1.3.1_
`interpret`-Example: https://codesandbox.io/s/xstateinitial-context-interpret-24bbz ⮕ this results in `context: undefined`
`useInterpret`-Example: https://codesandbox.io/s/xstateinitial-context-useinterpret-33950 ⮕ this results in `context: {}`

**Additional context**
- _There already was a little bit of discussion on discord about this:_
  - https://discord.com/channels/795785288994652170/799416943324823592/831217793479082024
  - https://discord.com/channels/795785288994652170/799416943324823592/831264646274482210
  - https://discord.com/channels/795785288994652170/799416943324823592/831444212322533418
  - https://discord.com/channels/795785288994652170/799416943324823592/831444929062240266
- Changing this behavior might be a breaking change ⚡.

### PR DESCRIPTION
**Pull request by VanTanev**
Fixes #2084

I set the changeset as minor bumps, because this could potentially break some existing code. Strictly, it's probably a major bump.

Also, I added `@testing-library/react-hooks` dependency because it made these tests much easier. However, I can remove it and re-implement with normal tests if that's @davidkpiano and @Andarist's preference.

### CODE REVIEW

**Comment by VanTanev** [comment]
cc @bemayr I had some free time and wrote this PR, hope that's OK.

**Comment by bemayr** [comment]
What a timing! 😂

File: .changeset/good-poets-sell.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+'@xstate/react': minor
+'@xstate/vue': minor
+---
+
+useInterpret()/useMachine() will no longer overwrite machine.context with an empty object when it's undefined.
```
Comment (davidkpiano) [inline .changeset/good-poets-sell.md:6]:
```suggestion
`useInterpret()` / `useMachine()` will no longer overwrite `machine.context` with an empty object when it's `undefined`.
```

**Review by davidkpiano — APPROVED** [review]

**Comment by davidkpiano** [comment]
> Also, I added `@testing-library/react-hooks` dependency because it made these tests much easier. However, I can remove it and re-implement with normal tests if that's @davidkpiano and @Andarist's preference.

That's fine! Wish I discovered it earlier, looks really useful!

**Comment by VanTanev** [comment]
@davidkpiano @Andarist This may need a major version bump, which is inconvenient for such a small change, but it may very well break machines defined without context and used in React.

If we're doing a major bump, I think this should throw: https://github.com/davidkpiano/xstate/blob/94c6d8eb47acb010c64c69a272d38a81e2d2b337/packages/core/src/utils.ts#L407

Or, alternatively - can we have an option that converts all `warn()` calls to throw? It seems that in almost all cases, `warn()` is used in places where you're doing something invalid, but the machine keeps on chugging. I'd rather get a hard error, instead of something that might slip past.

**Comment by Andarist** [comment]
> I added @testing-library/react-hooks dependency because it made these tests much easier.

Personally, I don't like this as much :P especially that we have other tests that don't use this and IMHO all tests should be just written using a single approach (what we have or this, whatever, just a single approach). What I like about current tests is that they resemble real usage and as such they are better usage examples/documentation.

**Comment by VanTanev** [comment]
@Andarist To test this without `@testing-library/react-hooks` we would need to do something like this: 
```typescript
  it('can extend the default context object', () => {
    const machine = createMachine({
      initial: 'foo',
      states: {
        foo: {}
      },
      context: {
        num: 1
      }
    });

    let result
    const App = () => {
      result = useInterpret(machine, { context: { num: 2 } })
      return null
    }
    
    render(<App />)

    expect(result.machine.context).toEqual({ num: 2 });
  });
```

Not sure if this resembles real usage in a meaningful way?

**Comment by Andarist** [comment]
If we consider the case that has been fixed here then this could be smth like this:
```js
  it('does not set a context object when there was none in the definition', () => {
    const machine = createMachine({
      initial: 'foo',
      states: {
        foo: {}
      },
    });

    const App = () => {
      const [state] = useMachine(machine)
      return <span>{state.context === undefined ? 1 : 0}</span>
    }
    
    const { container } = render(<App />)

    expect(container.textContent).toEqual('1');
  });
```

Other test cases would look similar.

**Comment by VanTanev** [comment]
@Andarist I see what you mean, but this feels like it's obfuscating the purpose of the test with unrelated code compared to the @testing-librart/react-hooks version that is directly testing the hook's return value.

**Comment by Andarist** [comment]
I just prefer not to test the hook's return value directly. Hooks are never used on their own so a test that uses them on their own is too close to being oriented on implementation details for me. 

This is totally a matter of opinion though so I'm not sure if it's worth debating that. I understand that my opinion might not be shared by others. There is still an issue that right now other tests are not written using this library and it's IMHO much better to keep a single way of testing things. If @davidkpiano would agree with you - then that would be fine with me but IMHO refactoring to this library should happen for all tests and should be done in a separate PR.

**Comment by davidkpiano** [comment]
> I just prefer not to test the hook's return value directly. Hooks are never used on their own so a test that uses them on their own is too close to being oriented on implementation details for me.
> 
> This is totally a matter of opinion though so I'm not sure if it's worth debating that. I understand that my opinion might not be shared by others. There is still an issue that right now other tests are not written using this library and it's IMHO much better to keep a single way of testing things. If @davidkpiano would agree with you - then that would be fine with me but IMHO refactoring to this library should happen for all tests and should be done in a separate PR.

Let's get this merged as-is, and discuss in a separate PR how we should unify the approaches. I see value in both having the same approach throughout all the tests, and the realization that testing UI is actually outside the scope of the hooks provided by `@xstate/react`, so it might actually be beneficial to use a hooks-only testing approach (especially if it makes things similar)

**Comment by VanTanev** [comment]
As per https://github.com/davidkpiano/xstate/pull/2100#issuecomment-821795396, I have incorporated the tests from #2100 into this PR. Thanks, @bemayr!

**Review by davidkpiano — APPROVED** [review]

**Comment by bemayr** [comment]
Thanks a lot @VanTanev for incorporating the changes/tests!

Regarding https://github.com/davidkpiano/xstate/pull/2100#issuecomment-821795396 and the problem of this being a breaking change @Andarist mentioned
> > @bemayr: So this should be postponed until new major releases of @xstate/react, @xstate/vue and @xstate/svelte?

> @Andarist: that would be my preferred approach here - u could target the next branch with this

should this PR target *next* now and be a major upgrade?

And regarding @xstate/svelte, the same problem exists there in `useMachine` as well, I wanted to fix this as well, but didn't have the time to dig into writing tests for this in svelte. Because this is an issue with consistency, should this especially be fixed across all frameworks at once?

**Review by bemayr — CHANGES_REQUESTED** [review]
Sorry to bother here again... 😅

File: packages/xstate-vue/test/useInterpret.test.ts [hunk 2]
```diff
@@ -18,4 +21,29 @@ describe('useInterpret composable function', () => {
...
+
+    const interpretService = interpret(machine);
+
+    const App = defineComponent({
+      setup() {
+        const useInterpretService = useInterpret(machine);
+
+        expect(useInterpretService.initialState.context).toEqual(
+          interpretService.initialState.context
+        );
```
[the remark this answers is not in the document]
Comment (VanTanev) [reply packages/xstate-vue/test/useInterpret.test.ts:41]: `<service>.state.context` does not exist until you start the service.  Not sure if changing the test is necessary.

**Comment by VanTanev** [comment]
Added the change to svelte, but I honestly don't want to deal with creating a svelte test for this, as I'm really unfamiliar with the framework.

File: packages/xstate-vue/test/useInterpret.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bemayr) [reply packages/xstate-vue/test/useInterpret.test.ts:41]: Oh yes, my bad, thanks for the explanation! 👏

**Comment by bemayr** [comment]
> Added the change to svelte, but I honestly don't want to deal with creating a svelte test for this, as I'm really unfamiliar with the framework.

Been there as well 😉, I implemented a test for this in VanTanev/xstate#1, is this ok?

**Comment by VanTanev** [comment]
Thanks, will merge it after work!

**Comment by Andarist** [comment]
> should this PR target next now and be a major upgrade?

IMHO - yes.

**Comment by VanTanev** [comment]
@Andarist This PR targets `@xstate/react` `@xstate/svelte` and `@xstate/vue`, it does not target `xstate`. Do we need to tie it to xstate-next? Isn't it unrelated?

**Comment by VanTanev** [comment]
This should be good for review

**Comment by davidkpiano** [comment]
> @Andarist This PR targets `@xstate/react` `@xstate/svelte` and `@xstate/vue`, it does not target `xstate`. Do we need to tie it to xstate-next? Isn't it unrelated?

If it only targets those, it doesn't need to be tied to `next` IMO, though I might be missing something @Andarist?

**Review by bemayr — APPROVED** [review]
Sorry, forgot that I had an open "Requested changes" here 👌

**Comment by VanTanev** [comment]
I went over the code once more, and am happy with its current state. @davidkpiano @Andarist could you review?
To re-iterate, this bumps `@xstate/react`, `@xstate/svelte` and `@xstate/vue` by a major version.

**Comment by davidkpiano** [comment]
I'm hesitant to merge this because it's a major change for an edge-case. We might want to sync it with v5 releases since that will also be a major release (and we can fix this for that as well).

Does it invalidate any tests? If not, maybe this doesn't need to be classified as a major change IMO.

**Comment by VanTanev** [comment]
@davidkpiano No, no previous test caught this behavior. I added new tests specifically to validate it.

**Comment by davidkpiano** [comment]
> @davidkpiano No, no previous test caught this behavior. I added new tests specifically to validate it.

@Andarist Given this, is a major change justified? Do we expect any major breakage here?

My hypothesis is that apps that didn't define `context` in the first place don't read from `state.context` at all, so they would never run into the problem of expecting a useless empty object and receiving `undefined`.

**Comment by Andarist** [comment]
@davidkpiano such code would suddenly throw at runtime: https://codesandbox.io/s/laughing-cdn-n2fb1?file=/src/App.js . Even if it could be improved (structure-wise) I feel like it's not unlikely that something like this (or similar) exists in production codebases.

OTOH, the discrepancy fixed by this PR is rather minor and harmless (IMHO). Yes, it would be great to have this consistent but for a case like this, I would prefer to keep this inconsistent for a while and fix it later.

I would have to recheck it but is it even possible to have undefined/null context in v5? I recall that you were changing this in some PR 🤔

**Comment by VanTanev** [comment]
@Andarist incorrect, it doesn't throw. You just get a warning in console: `Warning: Attempting to update undefined context` and it silently fails to update the context.

This is probably worse than throwing, and that warning should probably be converted to an exception.

**Comment by davidkpiano** [comment]
> I would have to recheck it but is it even possible to have undefined/null context in v5? I recall that you were changing this in some PR 🤔

Especially for strongly typing, I think `context` should always be defined; I'll make sure it is in v5.

@VanTanev If this is a major change, let's wait until we get a v5 alpha to time it with that. If this is only a minor change, we can release it sooner.

**Comment by Andarist** [comment]
@VanTanev incorrect, it throws 😉 

It doesn't throw within `assign` but it throws when rendering the `state.context.data` based on the initial state: https://codesandbox.io/s/youthful-haibt-mivb0?file=/src/App.js

**Comment by VanTanev** [comment]
@Andarist I see what you meant now. I was focusing on XState behavior :)

I'm fine with this being delayed until V5. It's not really tied to V5, but this seems to be the consensus, and the inconsistency is fairly minor.

**Comment by davidkpiano** [comment]
Closing as no longer valid in `@xstate/react@beta`
